File: ruby/051-resistor-color-trio/resistor_color_trio.rb
class ResistorColorTrio
  attr_reader :colors

  COLOR_BANDS = {
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }.freeze

  def initialize(colors)
    @colors = colors
  end

  def label
    sum = 0
    color1, color2, color3 = colors
    multiplier = COLOR_BANDS[color3.to_sym]
    sum += COLOR_BANDS[color1.to_sym] * 10**(multiplier + 1) + COLOR_BANDS[color2.to_sym] * 10**multiplier
    return "Resistor value: #{sum} ohms" if sum < 1000

    "Resistor value: #{sum/1000} kiloohms"
  end
end

File: go/085-resistor-color-trio/resistor_color_trio.go
package resistorcolortrio

import (
	"fmt"
	"math"
)

// Label describes the resistance value given the colors of a resistor.
// The label is a string with a resistance value with an unit appended
// (e.g. "33 ohms", "470 kiloohms").
func Label(colors []string) string {
	COLOR_CODE := map[string]int{
		"black":  0,
		"brown":  1,
		"red":    2,
		"orange": 3,
		"yellow": 4,
		"green":  5,
		"blue":   6,
		"violet": 7,
		"grey":   8,
		"white":  9,
	}

	val := (COLOR_CODE[colors[0]]*10 + COLOR_CODE[colors[1]]) * int(
		math.Pow(
			10,
			float64(COLOR_CODE[colors[2]]),
		),
	)

	if val >= 1000000000 {
		return fmt.Sprintf("%d gigaohms", val/1000000000)
	}

	if val >= 1000000 {
		return fmt.Sprintf("%d megaohms", val/1000000)
	}

	if val >= 1000 {
		return fmt.Sprintf("%d kiloohms", val/1000)
	}

	return fmt.Sprintf("%d ohms", val)
}
